### DISCUSSION TITLE
Convert Cart to Order: Override Stock Check?

### DISCUSSION THREAD

**Discussion by osseonews** [swellstores/community#41] [Help] [answered]
Your API has a functionality to convert a cart to an order (i.e. https://developers.swell.is/backend-api/carts/create-a-cart#convert-carts-to-orders). However, the conversion will fail if the product stock is unavailable. Is there a way to override this? In other words, we want the cart to convert even if the stock is unavailable. Of course, I know we can eliminate inventory tracking for the cart item on the backend, but we don't want to do that. For various scenarios, we need to both have inventory tracking on the backend, but also allow orders to go thru upon a cart conversion, even if the product stock is unavailable. Is there some argument we can send to the Convert to Cart api endpoint to force the order thru even if there is not enough stock of a product where we are tracking inventory? Thanks.

**Comment by logeshswell** [comment, marked as answer]
@osseonews That's a great question!
Yes, you can definitely do it. You're just one step away from enabling the `Allow purchases when out of stock` toggle under **Inventory** tab in Product details page.

Documentation Reference:
https://www.swell.is/help/products/product-management/product-inventory#purchasing-while-out-of-stock

**Reply by osseonews** [reply]
Thanks, but we know we can do this via a setting in the backend, but we don't want to do this on a general level for the whole store. What we are inquiring is how to do this on a one-off type basis via the API. Is there anything we can pass in the API that will override the failure when something is out of stock? How do we force the order to convert from the cart even if there is not enough stock? Is there some sort of flag in the API to force an order thru?

**Reply by logeshswell** [reply]
No @osseonews, we don't support API-level overriding the `stock_purchasable` flag for item at the moment, as we didn't find a proper use case for it.

**Reply by osseonews** [reply]
OK, np. BTW, the use case is that in reality when you run an e-commerce website with alot of SKU's, you want to have backorders for some items, but not others. This is because you have delivery schedules for certain items, and for an item that maybe delivered tomorrow, you may want to allow backorders, but for other items which may not get back in stock for awhile, you may want to not have backorders. This can change very quickly due to supplier delivery schedules. So having this as a general setting across the entire store ("Allow purchases when out of stock"), is not actually practical for a real business when different products are constantly going in and out of stock and supplier deliveries are changing every day (particularly in today's world).  You would want the ability to override this in specific cases and the easy way to do this would be via the API. So you could say, if such and such items are in the cart, override the purchase out of stock, and force the order thru. hope that makes sense.

**Reply by ericingram** [reply]
It's a good idea!

There is one workaround actually. The stock level is checked only when using the cart -> order conversion approach, where you create an order with cart_id=X, but you could create the order directly from cart details to get around that. The feature does a simple mapping from cart to order on the following fields (currently anyway):

```
const cartConversionFields = [
  'test',
  'items',
  'billing',
  'shipping',
  'shipment_rating',
  'shipment_discount',
  'schedule',
  'coupon_code',
  'coupon_id',
  'discounts',
  'taxes',
  'item_tax_included',
  'shipment_tax',
  'shipment_tax_included',
  'promotion_ids',
  'account_id',
  'account_logged_in',
  'account_info_saved',
  'account_credit_applied',
  'account_credit_amount',
  'giftcards',
  'currency',
  'display_currency',
  'display_locale',
  'notes',
  'comments',
  'gift',
  'gift_message',
  'metadata',
];
...

// Map fields from cart to order
for (const field of cartConversionFields) {
  if (order[field] === undefined) {
    order[field] = cart[field];
  }
}
...
```

And then it maps the order ID back to the cart record among a few other things, that you may or may not care about:

```
await swell.put('/carts/{id}', {
  id: cart.id,
  order_id: order.id,
  currency_rate: order.currency_rate,
  active: false,
  abandoned: cart.abandoned ? false : undefined,
  $events: {
    type: ['cart.updated', 'cart.converted'],
    data: {
      order_id: order.id,
    },
  },
});
```

**Comment by osseonews** [comment]
Just an aside, but the "Allow purchases when out of stock" is set on an individual product level! This means that to allow backorders we need to go into every single product we want to allow backorders for and toggle this to true. This is impractical. No chance, we are going to update many products like this and risk breaking something. you need to have a global setting: "Allow purchases when out of stock.", if you are going to keep this part of your API which blocks orders on an inventory issue. In any case, your existing inventory control really doesn't make any sense for any practical use case. There is no reason for an order to fail at the very last step when converting a cart to an order b/c of a lack of inventory. you would check inventory prior to converting to an order, not after! It is nuts that a customer can get to the final step in checkout and the order will fail b/c of this (unless of course we go into every single product and change the setting -  not happening). So now as it is, we need to manually create an order from a cart, per Eric's comment above, which is a huge, huge pain and prone to error.

Please just remove this feature from the cart API, convert cart to orders, or make it a Boolean type of API - so if it is true, the order will fail for inventory, and if false the order will succeed. At least then it can be overwritten at the API level. Honestly, I've done e-commerce for 15+ yrs and never once have I seen an order API endpoint like this. It is totally not based on any actual use case. The inventory is always checked prior to converting to an order (in cart) or there is a global setting for checking on the order for something like this. In your case, there is neither. You need to set this on every single product individually.

**Comment by osseonews** [comment]
@ericingram Can you please paste the full "Cart to Order" mapping function here that is internal to the API, so we don't need to reinvent the wheel and code our own "cart to order" in an attempt to override the inventory issue? Your code above has alot of ..., and we don't have access the code as it's not in the actual API packages. Thanks.

**Reply by ericingram** [reply]
@osseonews you are making a lot of  absolute statements without clear knowledge of the trade-offs, it seems. Not very productive.

Got your feedback, but you should re-phrase the question or request.
